File: javascript/json-transform/src/functions/common/ArgumentType.ts
import { ArgType } from "./ArgType";
import { BigDecimal } from "./FunctionHelpers";

export type ArgumentType = {
  type: ArgType; // ArgType
  position?: number; // default -1

  defaultIsNull?: boolean; // default false
  defaultBoolean?: boolean; // default false
  defaultString?: string; // default ""
  defaultEnum?: string; // default ""
  defaultInteger?: number; // default -1
  defaultLong?: number; // default -1L
  defaultBigDecimal?: number | string; // default -1

  aliases?: string[]; // default {};
};

File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/annotations/ArgumentType.java
package co.nlighten.jsontransform.functions.annotations;


import co.nlighten.jsontransform.functions.common.ArgType;

import java.lang.annotation.*;

/**
 * Describes a function argument.
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
@Repeatable(ArgumentsTypes.class)
public @interface ArgumentType {
    String value();
    ArgType type();
    int position() default -1;

    boolean defaultIsNull() default false;
    boolean defaultBoolean() default false;
    String defaultString() default "";
    String defaultEnum() default "";
    int defaultInteger() default -1;
    long defaultLong() default -1;
    double defaultBigDecimal() default -1;

    String[] aliases() default {};
}
